src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp

```diff
@@ -27,55 +26,56 @@ void characteristic_speeds(
     const Scalar<DataVector>& sound_speed_squared,
     const tnsr::i<DataVector, Dim>& normal) noexcept {
   const size_t num_grid_points = get<0>(shift).size();
-  constexpr size_t num_vectors = 5;
   // Allocating a single large buffer is much faster than many small buffers
-  std::unique_ptr<double[], decltype(&free)> temp_vectors(
-      static_cast<double*>(
-          malloc(num_grid_points * sizeof(double) * num_vectors)),
-      &free);
+  Variables<
+      tmpl::list<Tags::TempScalar<0>, Tags::TempScalar<1>, Tags::TempScalar<2>,
+                 Tags::TempScalar<3>, Tags::TempScalar<4>>>
+      temp_tensors{num_grid_points};
 
   // Because we don't require char_speeds to be of the correct size we use a
   // temp buffer for the dot product, then multiply by -1 assigning the result
   // to char_speeds.
-  Scalar<DataVector> normal_velocity{temp_vectors.get(), num_grid_points};
-  dot_product(make_not_null(&normal_velocity), normal, shift);
-  (*char_speeds)[0] = -1.0 * get(normal_velocity);
-  // Dim-fold degenerate eigenvalue
+  {
+    Scalar<DataVector>& normal_shift =
+        get<Tags::TempScalar<0>>(temp_tensors);
+    dot_product(make_not_null(&normal_shift), normal, shift);
+    (*char_speeds)[0] = -1.0 * get(normal_shift);
+  }
+  // Dim-fold degenerate eigenvalue, reuse normal_shift allocation
+  Scalar<DataVector>& normal_velocity =
+      get<Tags::TempScalar<0>>(temp_tensors);
   dot_product(make_not_null(&normal_velocity), normal, spatial_velocity);
-  // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
-  DataVector lapse_times_normal_velocity(temp_vectors.get() + num_grid_points,
-                                         num_grid_points);
-  lapse_times_normal_velocity = get(lapse) * get(normal_velocity);
-  (*char_speeds)[1] = (*char_speeds)[0] + lapse_times_normal_velocity;
+  Scalar<DataVector>& lapse_times_normal_velocity =
+      get<Tags::TempScalar<1>>(temp_tensors);
+  get(lapse_times_normal_velocity) = get(lapse) * get(normal_velocity);
+  (*char_speeds)[1] = (*char_speeds)[0] + get(lapse_times_normal_velocity);
```

Comment:
Is it necessary to make the `lapse_times_normal_velocity` reference? I don't think you use the value further below (you reuse the allocation but aren't reading from it) in which case you could just inline `speed[1] = speed[0] + lapse * normal_velocity`. Is there an advantage to separating the multiplication?